bus-mapping/src/circuit_input_builder.rs

```diff
@@ -561,61 +716,279 @@ impl<'a> CircuitInputStateRef<'a> {
             CallKind::Create | CallKind::Create2 => {
                 let init_code = get_create_init_code(step)?;
                 let code_hash = self.code_db.insert(init_code.to_vec());
                 (CodeSource::Memory, code_hash)
             }
             _ => {
-                let (found, account) = self.sdb.get_account(&address);
+                let code_address = match kind {
+                    CallKind::CallCode | CallKind::DelegateCall => {
+                        step.stack.nth_last(1)?.to_address()
+                    }
+                    _ => address,
+                };
+                let (found, account) = self.sdb.get_account(&code_address);
                 if !found {
-                    return Err(Error::AccountNotFound(address));
+                    return Err(Error::AccountNotFound(code_address));
                 }
-                (CodeSource::Address(address), (account.code_hash))
+                (CodeSource::Address(code_address), account.code_hash)
+            }
+        };
+
+        let get_memory_offset_length = |nth: usize| -> std::result::Result<_, Error> {
+            let offset = step.stack.nth_last(nth)?;
+            let length = step.stack.nth_last(nth + 1)?;
+            if length.is_zero() {
+                return Ok((0, 0));
             }
+            Ok((offset.low_u64(), length.low_u64()))
         };
-        self.push_call(kind, address, code_source, code_hash);
+        let (call_data_offset, call_data_length, return_data_offset, return_data_length) =
+            match kind {
+                CallKind::Call | CallKind::CallCode => {
+                    let call_data = get_memory_offset_length(3)?;
+                    let return_data = get_memory_offset_length(5)?;
+                    (call_data.0, call_data.1, return_data.0, return_data.1)
+                }
+                CallKind::DelegateCall | CallKind::StaticCall => {
+                    let call_data = get_memory_offset_length(2)?;
+                    let return_data = get_memory_offset_length(4)?;
+                    (call_data.0, call_data.1, return_data.0, return_data.1)
+                }
+                CallKind::Create | CallKind::Create2 => (0, 0, 0, 0),
+            };
+
+        let caller = &self.tx.calls[self.tx_ctx.call_index()];
+        let call = Call {
+            call_id: self.block_ctx.rwc.0,
+            caller_id: caller.call_id,
+            kind,
+            is_static: kind == CallKind::StaticCall || caller.is_static,
+            is_root: false,
+            is_persistent: caller.is_persistent && is_success,
+            is_success,
+            rw_counter_end_of_reversion: 0,
+            caller_address,
+            address,
+            code_source,
+            code_hash,
+            depth: caller.depth + 1,
+            value,
+            call_data_offset,
+            call_data_length,
+            return_data_offset,
+            return_data_length,
+        };
```

Comment:
But if with a new function, the argument will just be like almost all of fields, I don't think it's easier to read, especially when we can't get hint of argument's name.

Reply:
I'm also in favor of not adding a `new` function to build types that have so many fields.  I think it becomes harder to figure out what's going on because you'd see many `true, false, true, 1, 0, 1`.
Also, another reason is that `Call` is being used more like a struct and not a class, so I think not having a constructor is OK.

Reply:
You're right! I did not think about this. Resolving this.